hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/HoodieSparkMergeHandleV2.java

```diff
@@ -0,0 +1,410 @@
+
+      // Create the writer for writing the new version file
+      fileWriter = HoodieFileWriterFactory.getFileWriter(instantTime, newFilePath, hoodieTable.getStorage(),
+          config, writeSchemaWithMetaFields, taskContextSupplier, HoodieRecord.HoodieRecordType.SPARK);
+    } catch (IOException io) {
+      LOG.error("Error in update task at commit " + instantTime, io);
+      writeStatus.setGlobalError(io);
+      throw new HoodieUpsertException("Failed to initialize HoodieUpdateHandle for FileId: " + fileId + " on commit "
+          + instantTime + " on path " + hoodieTable.getMetaClient().getBasePath(), io);
+    }
+  }
+
+  protected void setWriteStatusPath() {
+    writeStatus.getStat().setPath(new StoragePath(config.getBasePath()), newFilePath);
+  }
+
+  protected void makeOldAndNewFilePaths(String partitionPath, String oldFileName, String newFileName) {
+    oldFilePath = makeNewFilePath(partitionPath, oldFileName);
+    newFilePath = makeNewFilePath(partitionPath, newFileName);
+  }
+
+  // TODO(yihua): is this still needed
+  /*
+  if (needsUpdateLocation()) {
+        record.unseal();
+        record.setNewLocation(new HoodieRecordLocation(instantTime, fileId));
+        record.seal();
+      }
+   */
+
+  public void write() {
+    boolean usePosition = config.getBooleanOrDefault(MERGE_USE_RECORD_POSITIONS);
+    Schema readerSchema;
+    Option<InternalSchema> internalSchemaOption = Option.empty();
+    if (!StringUtils.isNullOrEmpty(config.getInternalSchema())) {
+      readerSchema = HoodieAvroUtils.addMetadataFields(
+          new Schema.Parser().parse(config.getSchema()), config.allowOperationMetadataField());
+      internalSchemaOption = SerDeHelper.fromJson(config.getInternalSchema());
+    } else {
+      readerSchema = HoodieAvroUtils.addMetadataFields(
+          new Schema.Parser().parse(config.getSchema()), config.allowOperationMetadataField());
+    }
+    // TODO(yihua): reader schema is good enough for writer?
+    HoodieFileGroupReader<T> fileGroupReader = new HoodieFileGroupReader<>(
+        readerContext,
+        storage,
+        config.getBasePath(),
+        instantTime,
+        fileSlice,
+        readerSchema,
+        readerSchema,
+        internalSchemaOption,
+        hoodieTable.getMetaClient(),
+        hoodieTable.getMetaClient().getTableConfig().getProps(),
+        0,
+        Long.MAX_VALUE,
+        usePosition);
+    try {
+      fileGroupReader.initRecordIterators();
+      HoodieFileGroupReaderIterator<InternalRow> recordIterator
+          = (HoodieFileGroupReaderIterator<InternalRow>) fileGroupReader.getClosableIterator();
+      StructType sparkSchema = AvroConversionUtils.convertAvroSchemaToStructType(readerSchema);
+      while (recordIterator.hasNext()) {
+        InternalRow row = recordIterator.next();
+        HoodieKey recordKey = new HoodieKey(
+            row.getString(HoodieRecord.COMMIT_SEQNO_METADATA_FIELD_ORD),
```

Comment:
record key

Reply:
Fixed.  The code autocomplete gives me the wrong name.